### ISSUE TITLE
Carving with 2D data broken

### ISSUE THREAD

**Issue by k-dominik** [ilastik/ilastik#2109]
## Describe the bug
Carving fails during preprocessing with 2D data

## Expected behavior
Should just work (and it worked before, too)

## To Reproduce
Within the Carving workflow
1. Add some 2d single channel data
2. Go to the preprocessing applet, Click on _run_ -> :boom: 

## Error message/traceback
```pytb
ERROR 2019-09-17 11:30:37,628 log_exception 1077 140229934507840 Traceback (most recent call last):
  File ".../miniconda3/envs/idev/ilastik-meta/lazyflow/lazyflow/request/request.py", line 347, in _execute
    self._result = self.fn()
  File ".../miniconda3/envs/idev/ilastik-meta/lazyflow/lazyflow/slot.py", line 921, in __call__
    result_op = self.operator.call_execute(self.slot.top_level_slot, self.slot.subindex, self.roi, destination)
  File ".../miniconda3/envs/idev/ilastik-meta/lazyflow/lazyflow/operator.py", line 591, in call_execute
    return self.execute(slot, subindex, roi, result, **kwargs)
  File ".../sources/ilastik-meta/ilastik/ilastik/workflows/carving/opPreprocessing.py", line 411, in execute
    mst = self._opMstProvider.MST.value
  File ".../miniconda3/envs/idev/ilastik-meta/lazyflow/lazyflow/slot.py", line 1070, in value
    temp = self[:].wait()
  File ".../miniconda3/envs/idev/ilastik-meta/lazyflow/lazyflow/request/request.py", line 528, in wait
    return self._wait(timeout)
  File ".../miniconda3/envs/idev/ilastik-meta/lazyflow/lazyflow/request/request.py", line 556, in _wait
    self._wait_within_request(current_request)
  File ".../miniconda3/envs/idev/ilastik-meta/lazyflow/lazyflow/request/request.py", line 685, in _wait_within_request
    raise_with_traceback(exc_value, exc_tb)
  File ".../miniconda3/envs/idev/lib/python3.7/site-packages/future/utils/__init__.py", line 421, in raise_with_traceback
    raise exc.with_traceback(traceback)
  File ".../miniconda3/envs/idev/ilastik-meta/lazyflow/lazyflow/request/request.py", line 347, in _execute
    self._result = self.fn()
  File ".../miniconda3/envs/idev/ilastik-meta/lazyflow/lazyflow/slot.py", line 921, in __call__
    result_op = self.operator.call_execute(self.slot.top_level_slot, self.slot.subindex, self.roi, destination)
  File ".../miniconda3/envs/idev/ilastik-meta/lazyflow/lazyflow/operator.py", line 591, in call_execute
    return self.execute(slot, subindex, roi, result, **kwargs)
  File ".../sources/ilastik-meta/ilastik/ilastik/workflows/carving/opPreprocessing.py", line 246, in execute
    labelVolume = self.LabelImage(*roiFromShape(self.LabelImage.meta.shape)).wait()
  File ".../miniconda3/envs/idev/ilastik-meta/lazyflow/lazyflow/request/request.py", line 528, in wait
    return self._wait(timeout)
  File ".../miniconda3/envs/idev/ilastik-meta/lazyflow/lazyflow/request/request.py", line 556, in _wait
    self._wait_within_request(current_request)
  File ".../miniconda3/envs/idev/ilastik-meta/lazyflow/lazyflow/request/request.py", line 685, in _wait_within_request
    raise_with_traceback(exc_value, exc_tb)
  File ".../miniconda3/envs/idev/lib/python3.7/site-packages/future/utils/__init__.py", line 421, in raise_with_traceback
    raise exc.with_traceback(traceback)
  File ".../miniconda3/envs/idev/ilastik-meta/lazyflow/lazyflow/request/request.py", line 347, in _execute
    self._result = self.fn()
  File ".../miniconda3/envs/idev/ilastik-meta/lazyflow/lazyflow/slot.py", line 921, in __call__
    result_op = self.operator.call_execute(self.slot.top_level_slot, self.slot.subindex, self.roi, destination)
  File ".../miniconda3/envs/idev/ilastik-meta/lazyflow/lazyflow/operator.py", line 591, in call_execute
    return self.execute(slot, subindex, roi, result, **kwargs)
  File ".../miniconda3/envs/idev/ilastik-meta/lazyflow/lazyflow/operators/opUnblockedArrayCache.py", line 87, in execute
    self._execute_Output(slot, subindex, roi, result)
  File ".../miniconda3/envs/idev/ilastik-meta/lazyflow/lazyflow/operators/opSimpleBlockedArrayCache.py", line 106, in _execute_Output
    pool.wait()
  File ".../miniconda3/envs/idev/ilastik-meta/lazyflow/lazyflow/request/request.py", line 1360, in wait
    self._clear_finishing_requests()
  File ".../miniconda3/envs/idev/ilastik-meta/lazyflow/lazyflow/request/request.py", line 1436, in _clear_finishing_requests
    req.block()
  File ".../miniconda3/envs/idev/ilastik-meta/lazyflow/lazyflow/request/request.py", line 534, in block
    self._wait(timeout)  # No return value. Use wait()
  File ".../miniconda3/envs/idev/ilastik-meta/lazyflow/lazyflow/request/request.py", line 556, in _wait
    self._wait_within_request(current_request)
  File ".../miniconda3/envs/idev/ilastik-meta/lazyflow/lazyflow/request/request.py", line 641, in _wait_within_request
    raise_with_traceback(exc_type(exc_value), exc_tb)
  File ".../miniconda3/envs/idev/lib/python3.7/site-packages/future/utils/__init__.py", line 421, in raise_with_traceback
    raise exc.with_traceback(traceback)
  File ".../miniconda3/envs/idev/ilastik-meta/lazyflow/lazyflow/request/request.py", line 347, in _execute
    self._result = self.fn()
  File ".../miniconda3/envs/idev/ilastik-meta/lazyflow/lazyflow/operators/opSimpleBlockedArrayCache.py", line 81, in copy_block
    self._execute_Output_impl(clipped_block_roi, result[roiToSlice(*output_roi)])
  File ".../miniconda3/envs/idev/ilastik-meta/lazyflow/lazyflow/operators/opUnblockedArrayCache.py", line 112, in _execute_Output_impl
    self._fetch_and_store_block(request_roi, out=result)
  File ".../miniconda3/envs/idev/ilastik-meta/lazyflow/lazyflow/operators/opUnblockedArrayCache.py", line 150, in _fetch_and_store_block
    block_data = req.wait()
  File ".../miniconda3/envs/idev/ilastik-meta/lazyflow/lazyflow/request/request.py", line 528, in wait
    return self._wait(timeout)
  File ".../miniconda3/envs/idev/ilastik-meta/lazyflow/lazyflow/request/request.py", line 556, in _wait
    self._wait_within_request(current_request)
  File ".../miniconda3/envs/idev/ilastik-meta/lazyflow/lazyflow/request/request.py", line 685, in _wait_within_request
    raise_with_traceback(exc_value, exc_tb)
  File ".../miniconda3/envs/idev/lib/python3.7/site-packages/future/utils/__init__.py", line 421, in raise_with_traceback
    raise exc.with_traceback(traceback)
  File ".../miniconda3/envs/idev/ilastik-meta/lazyflow/lazyflow/request/request.py", line 347, in _execute
    self._result = self.fn()
  File ".../miniconda3/envs/idev/ilastik-meta/lazyflow/lazyflow/request/request.py", line 828, in __call__
    return self.func(*totalargs, **self.kwargs)
  File ".../miniconda3/envs/idev/ilastik-meta/lazyflow/lazyflow/slot.py", line 921, in __call__
    result_op = self.operator.call_execute(self.slot.top_level_slot, self.slot.subindex, self.roi, destination)
  File ".../miniconda3/envs/idev/ilastik-meta/lazyflow/lazyflow/operator.py", line 591, in call_execute
    return self.execute(slot, subindex, roi, result, **kwargs)
  File ".../miniconda3/envs/idev/ilastik-meta/lazyflow/lazyflow/operators/opCacheFixer.py", line 65, in execute
    self.Input(roi.start, roi.stop).writeInto(result).wait()
  File ".../miniconda3/envs/idev/ilastik-meta/lazyflow/lazyflow/request/request.py", line 528, in wait
    return self._wait(timeout)
  File ".../miniconda3/envs/idev/ilastik-meta/lazyflow/lazyflow/request/request.py", line 556, in _wait
    self._wait_within_request(current_request)
  File ".../miniconda3/envs/idev/ilastik-meta/lazyflow/lazyflow/request/request.py", line 685, in _wait_within_request
    raise_with_traceback(exc_value, exc_tb)
  File ".../miniconda3/envs/idev/lib/python3.7/site-packages/future/utils/__init__.py", line 421, in raise_with_traceback
    raise exc.with_traceback(traceback)
  File ".../miniconda3/envs/idev/ilastik-meta/lazyflow/lazyflow/request/request.py", line 347, in _execute
    self._result = self.fn()
  File ".../miniconda3/envs/idev/ilastik-meta/lazyflow/lazyflow/request/request.py", line 828, in __call__
    return self.func(*totalargs, **self.kwargs)
  File ".../miniconda3/envs/idev/ilastik-meta/lazyflow/lazyflow/slot.py", line 921, in __call__
    result_op = self.operator.call_execute(self.slot.top_level_slot, self.slot.subindex, self.roi, destination)
  File ".../miniconda3/envs/idev/ilastik-meta/lazyflow/lazyflow/operator.py", line 591, in call_execute
    return self.execute(slot, subindex, roi, result, **kwargs)
  File ".../sources/ilastik-meta/ilastik/ilastik/workflows/carving/opPreprocessing.py", line 211, in execute
    reduce_to=self.ReduceTo.value,
ValueError: could not broadcast input array from shape (250,250) into shape (250,250,1)

```

## Desktop (please complete the following information):
 - ilastik version: 1.3.3b2

ilastik ilastik/ilastik@6b143f732f17bf013a84ccfca2577ce6c486f146
lazyflow ilastik/lazyflow@ae5a07647237f50590977f1b3d20cc4af6feef4e
volumina ilastik/volumina@7df7e08bee7d3ee1fcb9f67c8871714adeda2a2b

**Comment by constantinpape** [comment]
This probably happened when I refactored some of the carving pre-processing recently.
I think the issue is that `resultView[...]` has shape `(X, Y, 1)`, but `watershed_and_agglomeate` returns shape `(X, Y)`. See https://github.com/ilastik/ilastik/blob/master/ilastik/workflows/carving/opPreprocessing.py#L207-L212. 

I think that the easiest solution would be `resultView[..., 0]` or `resultView[:, :, 0]` if we want to be more explicit. 

It would be good to also test carving 2d [here](https://github.com/ilastik/ilastik/blob/master/tests/test_workflows/testCarvingGui.py), so we can be sure not to break this again.

**Comment by k-dominik** [comment]
thank you very much for checking that out @constantinpape,
Regarding tests, I'll try to think of something that doesn't result in code duplication

**Comment by emilmelnikov** [comment]
I tried to refactor tests in https://github.com/ilastik/ilastik/pull/2114, but it it seems that the new 2D test data is needed. Then we can define `class TestCarvingGui3D` with class-level constants pointing to the new test data files.

**Comment by constantinpape** [comment]
> I tried to refactor tests in #2114, but it it seems that the new 2D test data is needed.

Maybe we can just cat a slice from the 3d test data for this?

**Comment by k-dominik** [comment]
> > I tried to refactor tests in #2114, but it it seems that the new 2D test data is needed.
> 
> Maybe we can just cat a slice from the 3d test data for this?

sure, why not. One would have to adjust all the hard-coded numbers there then and generate new reference files for the 2d case.

**Comment by k-dominik** [comment]
fixed via #2114
